### ISSUE TITLE
Map Matching API Original Index

### ISSUE THREAD

**Issue by schettgen** [valhalla/valhalla#4738]
## TL;DR 

add options to have relation between map matching result from and original request lat lon's
Return values may also be changed to sth like this 
```  {
        "location":
          { "lat": -35.28459730000001, "lon": 149.1283703 },
  "originalIndex": 0,
... maybe more infos here ...
      },
```

### Preconditions

The following reqeust

```
{
	"shape": [

		{
			"lat": 50.164416381682194, 
			"lon": 7.197217592199645
		},

		{
			"lat": 50.163420049968515,
			"lon": 7.196542354443883
		},

		{
			"lat": 50.163565850347254, 
			"lon": 7.197999037816894
		}

	],
	"costing": "auto",
	"shape_match": "map_snap",
	"filters": {
		"attributes": [
			"edge.names",
			"edge.id",
			"edge.weighted_grade",
			"edge.speed"
		],
		"action": "include"
	},
	"trace_options": {
		"search_radius": 100,
		"debug": true
	}
}
```

Will return 
```
{
    "trip": {
        "locations": [
            {
                "type": "break",
                "lat": 50.164416,
                "lon": 7.197217,
                "original_index": 0
            },
            {
                "type": "break",
                "lat": 50.163565,
                "lon": 7.197999,
                "original_index": 2
            }
        ],
        "legs": [
            {
                "maneuvers": [
                    {
                        "type": 1,
                        "instruction": "Drive east on Im Weingarten.",
                        "verbal_succinct_transition_instruction": "Drive east. Then Turn right onto Kernstraße.",
                        "verbal_pre_transition_instruction": "Drive east on Im Weingarten. Then Turn right onto Kernstraße.",
                        "verbal_post_transition_instruction": "Continue for 30 meters.",
                        "street_names": [
                            "Im Weingarten"
                        ],
                        "time": 2.844,
                        "length": 0.027,
                        "cost": 2.56,
                        "begin_shape_index": 0,
                        "end_shape_index": 2,
                        "verbal_multi_cue": true,
                        "travel_mode": "drive",
                        "travel_type": "car"
                    },
                    {
                        "type": 10,
                        "instruction": "Turn right onto Kernstraße.",
                        "verbal_transition_alert_instruction": "Turn right onto Kernstraße.",
                        "verbal_succinct_transition_instruction": "Turn right.",
                        "verbal_pre_transition_instruction": "Turn right onto Kernstraße.",
                        "verbal_post_transition_instruction": "Continue for 100 meters.",
                        "street_names": [
                            "Kernstraße"
                        ],
                        "time": 23.305,
                        "length": 0.134,
                        "cost": 26.972,
                        "begin_shape_index": 2,
                        "end_shape_index": 7,
                        "travel_mode": "drive",
                        "travel_type": "car"
                    },
                    {
                        "type": 15,
                        "instruction": "Turn left onto Brühlstraße.",
                        "verbal_transition_alert_instruction": "Turn left onto Brühlstraße.",
                        "verbal_succinct_transition_instruction": "Turn left. Then, in 100 meters, You will arrive at your destination.",
                        "verbal_pre_transition_instruction": "Turn left onto Brühlstraße. Then, in 100 meters, You will arrive at your destination.",
                        "verbal_post_transition_instruction": "Continue for 100 meters.",
                        "street_names": [
                            "Brühlstraße"
                        ],
                        "time": 11.999,
                        "length": 0.107,
                        "cost": 19.435,
                        "begin_shape_index": 7,
                        "end_shape_index": 12,
                        "verbal_multi_cue": true,
                        "travel_mode": "drive",
                        "travel_type": "car"
                    },
                    {
                        "type": 4,
                        "instruction": "You have arrived at your destination.",
                        "verbal_transition_alert_instruction": "You will arrive at your destination.",
                        "verbal_pre_transition_instruction": "You have arrived at your destination.",
                        "time": 0.0,
                        "length": 0.0,
                        "cost": 0.0,
                        "begin_shape_index": 12,
                        "end_shape_index": 12,
                        "travel_mode": "drive",
                        "travel_type": "car"
                    }
                ],
                "summary": {
                    "has_time_restrictions": false,
                    "has_toll": false,
                    "has_highway": false,
                    "has_ferry": false,
                    "min_lat": 50.163342,
                    "min_lon": 7.196568,
                    "max_lat": 50.164421,
                    "max_lon": 7.198001,
                    "time": 38.149,
                    "length": 0.268,
                    "cost": 48.968
                },
                "shape": "ejxt~AgahvLc@iLj@sIdW~ZtLpNfHbIhFzFrF`BlAkDp@aEmHsr@yAiQm@cG"
            }
        ],
        "summary": {
            "has_time_restrictions": false,
            "has_toll": false,
            "has_highway": false,
            "has_ferry": false,
            "min_lat": 50.163342,
            "min_lon": 7.196568,
            "max_lat": 50.164421,
            "max_lon": 7.198001,
            "time": 38.149,
            "length": 0.268,
            "cost": 48.968
        },
        "status_message": "Found route between points",
        "status": 0,
        "units": "kilometers",
        "language": "en-US"
    }
}
```

### Issue
The response contains the shape als polyline. 
The problem here is that I don't have any relation to the original index of the positions from the request. 

Google for example will return 
```
{
  "snappedPoints":
    [
      {
        "location":
          { "latitude": -35.27800489993019, "longitude": 149.129531998742 },
        "originalIndex": 0,
        "placeId": "ChIJr_xl0GdNFmsRsUtUbW7qABM",
      },
      {
        "location":
          { "latitude": -35.2784195, "longitude": 149.12946589999999 },
        "placeId": "ChIJr_xl0GdNFmsRsUtUbW7qABM",
      },
    ...
      {
        "location":
          { "latitude": -35.28459730000001, "longitude": 149.1283703 },
        "placeId": "ChIJtWxAZmpNFmsRlbUCkc6VtnA",
      },
      {
        "location":
          { "latitude": -35.284728747199374, "longitude": 149.12834860726772 },
        "originalIndex": 7,
        "placeId": "ChIJtWxAZmpNFmsRlbUCkc6VtnA",
      },
    ],
}
```
So there the original index from the request is keept

### Why

We sample data from tracker which will sent us lat lon at a specific time (with some other infos).
We plan to use map matching to put a trip onto a map which exactlly is following the street. 

Since the tracker is sent more properties as well we want to display them somehow like this 
![image](https://github.com/valhalla/valhalla/assets/76427220/337b58b7-3718-4533-bcb5-58bb6bdc42bf)

We need to show speed, values of GPIO's and some other values as well in that tooltip. 
Therefore we need the original index reference to put that infos on the map where we sampled it.

**Comment by kevinkreiser** [comment]
ok by looking at this a little bit i think i understand that you are hitting the `trace_route` endpoint. however it seems what you are asking for us a way for you to attach information to the location object that just gets passed through? in other words its not for the router to actually make use of, you just want to get it back in the serialized location in the response, is that right?

if that makes sense then i can propose a way of doing that and we can support it for all endpoints that serialize a location object back out. let me know if i have understood you correctly.

**Comment by schettgen** [comment]
In the end I just need a reference to the original lat lon object from the request

e.g.   /trace_route
```
{
	"shape": [

		{
			"lat": 50.164416381682194, 
			"lon": 7.197217592199645
		},

		{
			"lat": 50.163420049968515,
			"lon": 7.196542354443883
		},

		{
			"lat": 50.163565850347254, 
			"lon": 7.197999037816894
		}

	],
....
}
```

should return sth like this (please ignore lat lon data since it's just a example and not logically correct) 

```
{
    "trip": {
        "locations": [
            {
                "type": "break",
                "lat": 50.164416,
                "lon": 7.197217,
                "original_index": 0
            },
            {
                "type": "break",
                "lat": 50.163565,
                "lon": 7.197999,
                "original_index": 2
            }
        ],
        "legs": [
            {
                "maneuvers": [ ...],
                "summary": {...},
                "shape":  // CHANGES HERE <<<<<<<<<<<<<<<
                 [
		{
			"lat": 50.164416381682194, 
			"lon": 7.197217592199645,
			"original_index": 0
		},

		{
			"lat": 50.163420049968515,
			"lon": 7.196542354443883 
//<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< no original index here since this position is a additional position to match a curve (for example) 
		},

		{
			"lat": 50.163565850347254, 
			"lon": 7.197999037816894,
			"original_index": 1
		}
....
		{
			"lat": 50.163565850347254, 
			"lon": 7.197999037816894,
			"original_index": 3
		}
....

	],

            }
        ],
        "summary": {... },
        "status_message": "Found route between points",
        "status": 0,
        "units": "kilometers",
        "language": "en-US"
    }
}
```

That would also make it more compatible to google's api and maybe help alot of people to switch away from google's snap to road


1 response lat lon object will have one original index id or zero
1 request lat lon object will have one reference pointing to it from the response lat lon object

the response lat lon object could have a original index but still have a different lat lng than the object where it's pointing to

That will make it possible to map our sample data to the corresponding point

**Comment by schettgen** [comment]
@kevinkreiser so in case in can support you somehow or if you still have questions just let me know :)
